packages/engine/vscode/src/lib/editor.ts

```diff
@@ -1 +1,84 @@
+import { CommandPlugin } from "./command";
+
+export interface EditorOptions extends PluginOptions {
+  language: string;
+}
+export class EditorPlugin extends CommandPlugin implements MethodApi<IEditor> {
+  private decoration: TextEditorDecorationType;
+  private diagnosticCollection: DiagnosticCollection;
+  constructor(options: EditorOptions) {
+    super(editorProfile);
+    super.setOptions(options);
+  }
+  setOptions(options: EditorOptions) {
+    super.setOptions(options);
+  }
+  onActivation() {
+    this.decoration = window.createTextEditorDecorationType({
+      backgroundColor: 'editor.lineHighlightBackground',
+      isWholeLine: true,
+    });
+    // @ts-ignore
+    this.diagnosticCollection = languages.createDiagnosticCollection(this.options.language);
+  }
+  onDeactivation() {
+    this.decoration.dispose();
+  }
+  async highlight(position: HighlightPosition, filePath: string, hexColor: string): Promise<void> {
+    const editors = window.visibleTextEditors;
+    // Parse `filePath` to ensure if a valid file path was supplied
+    const editor = editors.find(editor => editor.document.uri.path === Uri.parse(filePath).path);
+    if(editor) {
+      const start: Position = new Position(position.start.line, position.start.column);
+      const end: Position = new Position(position.end.line, position.end.column);
+      const newDecoration = { range: new Range(start, end) };
+      if (hexColor) {
+        this.decoration = window.createTextEditorDecorationType({
+          backgroundColor: hexColor,
+          isWholeLine: true,
+        });
+      }
+      return editor.setDecorations(this.decoration, [newDecoration]);
+    } else {
+      throw new Error(`Could not find file ${filePath}`);
+    }
+  }
+  async discardHighlight(): Promise<void> {
+    return this.decoration.dispose();
+  }
+  async discardHighlightAt(line: number, filePath: string): Promise<void> {
+    return this.decoration.dispose();
+  }
+  private getEditor(filePath?: string): TextEditor {
+    const editors = window.visibleTextEditors;
+    return filePath ? editors.find(editor => editor.document.uri.path === Uri.parse(filePath).path) : window.activeTextEditor
+  }
+  async addAnnotation(annotation: Annotation, filePath?: string): Promise<void> {
+    // This function should append to existing map
+    // Ref: https://code.visualstudio.com/api/language-extensions/programmatic-language-features#provide-diagnostics
+    // const fileUri = window.activeTextEditor ? window.activeTextEditor.document.uri : undefined; // TODO: we might want to supply path to addAnnotation function
+    const editor = this.getEditor(filePath);
+    const canonicalFile: string = editor.document.uri.fsPath;
+    const diagnosticMap: Map<string, Diagnostic[]> = new Map();
+    const range = new Range(annotation.row - 1, annotation.column, annotation.row - 1, annotation.column);
+    const diagnostics = diagnosticMap.get(canonicalFile) ? diagnosticMap.get(canonicalFile) : [];
```

Comment:
How is it possible that you get a diagnostic from the `diagnosticMap` ? You just initialize the map two lines above and you don't mutate it in-between.
The way I see it you should have : 
```typescript
const editor = this.getEditor(filePath);
const path = editor.document.uri.fsPath;
const range = new Range(annotation.row - 1, annotation.column, annotation.row - 1, annotation.column);
const severity = diagnosticSeverity[annotation.type];
const diagnostic = new Diagnostic(range, annotation.text, severity);
this.diagnosticCollection?.set(Uri.file(path), [ diagnostic ]);
```